pkg/util/pki/csr.go

```diff
@@ -415,17 +472,78 @@ func SignCertificate(template *x509.Certificate, issuerCert *x509.Certificate, p
 	return pemBytes.Bytes(), cert, err
 }
 
-// SignCSRTemplate signs a certificate template usually based upon a CSR. This
+func hasProhibitingPathLen(caCert *x509.Certificate, chainNumber int) bool {
+	if !caCert.BasicConstraintsValid {
+		// if there's no basicConstraint extension, then there's no pathLen
+		// and therefore no restriction caused by this cert
+		// (this likely wouldn't work in public TLS)
+		return false
+	}
+
+	if caCert.MaxPathLen == -1 || (caCert.MaxPathLen == 0 && !caCert.MaxPathLenZero) {
+		// MaxPathLen = 0 means no pathLen was set if MaxPathLenZero is false
+		// and so there's no restriction caused by this cert
+		// alternatively, MaxPathLen == -1 also means it was unset and so there's no
+		// restriction
+		return false
+	}
+
+	// if we're here, there's a valid pathLen set on the cert
+	caPathLen := caCert.MaxPathLen
+
+	// the cert that issued the leaf has chainNumber = 0, and so
+	//   needs a pathLen >= 1 to have an effectivePathLen over 0
+	// the cert that issued that issuer has chainNumber = 1, and so
+	//   needs a pathLen >= 2 to have an effectivePathLen over 0
+	// etc.
+
+	effectivePathLen := caPathLen - chainNumber
+
+	// if the effectivePathLen is less than or equal to zero, then issuing a
+	// CA certificate would violate the pathLen constraint for caCert
+
+	return effectivePathLen <= 0
+}
+
+// SignCSRTemplate signs a certificate template, which is usually based upon a CSR. This
 // function expects all fields to be present in the certificate template,
-// including it's public key.
-// It returns the PEM bundle containing certificate data and the CA data, encoded in PEM format.
+// including its public key.
+// It returns a PEM bundle containing certificate data along with any available chain
+// certificates which were passed in caCerts.
 func SignCSRTemplate(caCerts []*x509.Certificate, caKey crypto.Signer, template *x509.Certificate) (PEMBundle, error) {
 	if len(caCerts) == 0 {
 		return PEMBundle{}, errors.New("no CA certificates given to sign CSR template")
 	}
 
+	caCerts, err := OrderCertificateChain(caCerts)
+	if err != nil {
+		return PEMBundle{}, err
+	}
+
 	issuingCACert := caCerts[0]
 
+	validPrivateKey, err := PublicKeyMatchesCertificate(caKey.Public(), issuingCACert)
+	if err != nil {
+		return PEMBundle{}, fmt.Errorf("failed to validate that signer key matches issuer: %w", err)
+	}
+
```

Comment:
self-review note: it's worth noting here that this change in the "experimental" commit could be in theory be a breaking change; that is, ordering the issuers first  and then ensuring that caKey matches the issuer could break some strange, probably broken scenarios where people were manually editing, say, the `Secret` resource which held their CA issuer.

I think those environments for which this might be a breaking change are likely to be pretty broken anyway, so I'm not concerned about breaking them.

This could also break people who import cert-manager as a library, but we don't really support that in any case.